web/elm/src/Build/Build.elm

```diff
@@ -199,206 +224,249 @@ getUpdateMessage model =
+            ( model, effects ++ [ DoAbortBuild buildId ] )
 
         ToggleStep id ->
             updateOutput
-                (Build.Output.handleStepTreeMsg <| StepTree.toggleStep id)
-                model
+                (Build.Output.Output.handleStepTreeMsg <| StepTree.toggleStep id)
+                ( model, effects )
 
         SwitchTab id tab ->
             updateOutput
-                (Build.Output.handleStepTreeMsg <| StepTree.switchTab id tab)
-                model
+                (Build.Output.Output.handleStepTreeMsg <| StepTree.switchTab id tab)
+                ( model, effects )
 
         SetHighlight id line ->
             updateOutput
-                (Build.Output.handleStepTreeMsg <| StepTree.setHighlight id line)
-                model
+                (Build.Output.Output.handleStepTreeMsg <| StepTree.setHighlight id line)
+                ( model, effects )
 
         ExtendHighlight id line ->
             updateOutput
-                (Build.Output.handleStepTreeMsg <| StepTree.extendHighlight id line)
-                model
-
-        RevealCurrentBuildInHistory ->
-            ( model, [ Scroll ToCurrentBuild ] )
+                (Build.Output.Output.handleStepTreeMsg <| StepTree.extendHighlight id line)
+                ( model, effects )
 
         ScrollBuilds event ->
-            if event.deltaX == 0 then
-                ( model, [ Scroll (Builds event.deltaY) ] )
-
-            else
-                ( model, [ Scroll (Builds -event.deltaX) ] )
-
-        ClockTick now ->
             let
-                newModel =
-                    { model
-                        | now = Just now
-                        , hoveredCounter = model.hoveredCounter + 1
-                    }
-            in
-            updateOutput
-                (Build.Output.handleStepTreeMsg <|
-                    StepTree.updateTooltip newModel
-                )
-                newModel
-
-        WindowScrolled fromBottom ->
-            if fromBottom == 0 then
-                ( { model | autoScroll = True }, [] )
-
-            else
-                ( { model | autoScroll = False }, [] )
-
-        NavTo route ->
-            ( model, [ NavigateTo <| Routes.toString route ] )
-
-        NewCSRFToken token ->
-            ( { model | csrfToken = token }, [] )
-
-        KeyPressed keycode ->
-            handleKeyPressed (Char.fromCode keycode) model
+                scroll =
+                    if event.deltaX == 0 then
+                        [ Scroll (Element "builds" event.deltaY) ]
 
-        KeyUped keycode ->
-            case Char.fromCode keycode of
-                'T' ->
-                    ( { model | previousTriggerBuildByKey = False }, [] )
+                    else
+                        [ Scroll (Element "builds" -event.deltaX) ]
 
-                _ ->
-                    ( model, [] )
+                checkVisibility =
+                    case model.history |> List.reverse |> List.head of
```

Comment:
Same last here, and it's model.history again. We... don't have any easy way of holding a pointer to the last element, because immutable copying lists :( But probably there's *some* way of cleaning this up, even if it's just a `lastBuild : Model -> Maybe Build` utility in the file somewhere.